### ISSUE TITLE
Is there a plan to support QAT with conv_bn folding?

### ISSUE THREAD

**Issue by Zehaos** [openvinotoolkit/nncf_pytorch#81]
Conv_bn folding is mentioned in https://arxiv.org/pdf/1806.08342.pdf 3.2.2 for getting better QAT accuracy.
It has been implemented in pytorch (https://github.com/pytorch/pytorch/blob/master/torch/nn/intrinsic/qat/modules/conv_fused.py#L82-L92)

Would it be implemented in nncf? Thanks. (I think it is important for QAT from scratch.)

**Comment by vshampor** [comment]
Greetings, @Zehaos! 

As far as I understood from the code and the paper, at training time the batch norm folding boils down to not inserting quantization operations in-between `Conv` and the immediately following `BatchNorm` operations. This kind of behavior has always been present in NNCF, right from the start - it does not insert FakeQuantize operations for activation tensors in-between certain patterns, such as Conv + RELU, Conv + BN + RELU and others. Basically, in the default, non-HW-specific scenarion we analyze the model graph for [these](https://github.com/openvinotoolkit/nncf_pytorch/blob/be622d4a6011e328e04c2d40779297fae4e43c45/nncf/quantization/algo.py#L660) operation patterns, and quantize the largest subgraph match - this ensures that we obtain, for example, an `FQ -> Conv -> BN -> RELU -> FQ` final execution pattern instead of `FQ -> Conv -> FQ -> BN -> FQ -> RELU -> FQ`.

You can compare the first two pictures under [Exploring the compressed model](https://github.com/openvinotoolkit/nncf_pytorch/blob/develop/docs/Usage.md#exploring-the-compressed-model) for an illustration of the fact that, for instance, `Conv -> RELU` patterns are *not* broken down by FakeQuantize operations. The same happens in NNCF for the `Conv -> BN` and other related patterns.

**Comment by Zehaos** [comment]
@vshampor  Thanks for your reply! 

It is not about the insert point of fakequant op. The magic of conv_bn folding lies in fake-quantization of the the conv kernel (bn meged) during training.

As far as I know, NNCF only fake-quantize conv kernel (bn not merged). This is not consistance with the inference phase (bn merged), that will lead to accuracy drop.

**Comment by vshampor** [comment]
@Zehaos, after reviewing the code more carefully, I understood what you meant there. Indeed, NNCF currently does not support the BN folding in the way it is implemented in PyTorch. Note, however, that the PyTorch way is also not exactly the way proposed in https://arxiv.org/pdf/1806.08342.pdf. 

I think it is, in theory, possible to enable this with NNCF. The BN folding might help with low-batch training scenarios, such as object detection. However, we will have to do some experiments first to ensure that this feature actually improves general quantized model accuracy in a tangible way - this introduces computation overhead during training, after all. Will track this in #88.

**Comment by Zehaos** [comment]
Great!

**Comment by xiaoyaopeng** [comment]
I think BN folding can have a better speed during inference phase. NNCF will spend more instructions to calculate BN layer when inference on embeded devices. Am I right?

**Comment by AlexKoff88** [comment]
NNCF does not do inference on embedded devices. To be honest we use GPUs for training and validation (inference). The main reason why we need BN is that it helps to restore the accuracy and having a stable training when we go to aggressive optimizations, like Filter Pruning or Sparsity, when the model gets highly perturbed. But BNs are definitely being fused when we go to inference with OpenVINO. Hope it helps.
